File: go-aethermq/internal/broker/dlq.go
package broker

import (
	"sync"
	"time"

	"github.com/aethermq/aethermq/internal/common"
)

type DLQManager struct {
	mu      sync.RWMutex
	queues  map[string]*DLQueue
}

type DLQueue struct {
	topic      string
	group      string
	messages   []*common.Message
	maxSize    int
	retryCount map[string]int
}

func NewDLQManager() *DLQManager {
	return &DLQManager{
		queues: make(map[string]*DLQueue),
	}
}

func (m *DLQManager) AddToDLQ(msg *common.Message, group string, reason string) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	key := dlqKey(msg.Topic, group)
	queue, ok := m.queues[key]
	if !ok {
		queue = &DLQueue{
			topic:      msg.Topic,
			group:      group,
			messages:   make([]*common.Message, 0),
			maxSize:    10000,
			retryCount: make(map[string]int),
		}
		m.queues[key] = queue
	}

	if len(queue.messages) >= queue.maxSize {
		queue.messages = queue.messages[1:]
	}

	if msg.Headers == nil {
		msg.Headers = make(map[string]string)
	}
	msg.Headers["x-dlq-reason"] = reason
	msg.Headers["x-dlq-original-topic"] = msg.Topic
	msg.Headers["x-dlq-timestamp"] = time.Now().Format(time.RFC3339)

	queue.messages = append(queue.messages, msg)
	return nil
}

func (m *DLQManager) GetDLQMessages(topic, group string, offset, limit int) ([]*common.Message, int, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	key := dlqKey(topic, group)
	queue, ok := m.queues[key]
	if !ok {
		return nil, 0, nil
	}

	total := len(queue.messages)
	if offset >= total {
		return nil, total, nil
	}

	end := offset + limit
	if end > total {
		end = total
	}

	return queue.messages[offset:end], total, nil
}

func (m *DLQManager) RetryMessage(topic, group string, messageID string) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	key := dlqKey(topic, group)
	queue, ok := m.queues[key]
	if !ok {
		return nil
	}

	for i, msg := range queue.messages {
		if msg.MessageID == messageID {
			queue.messages = append(queue.messages[:i], queue.messages[i+1:]...)
			msg.DeliverAt = 0
			msg.RetryCount++
			return nil
		}
	}

	return nil
}

func (m *DLQManager) ListDLQs() []DLQInfo {
	m.mu.RLock()
	defer m.mu.RUnlock()

	var infos []DLQInfo
	for key, queue := range m.queues {
		infos = append(infos, DLQInfo{
			Key:       key,
			Topic:     queue.topic,
			Group:     queue.group,
			Count:     len(queue.messages),
			MaxSize:   queue.maxSize,
		})
	}
	return infos
}

type DLQInfo struct {
	Key     string `json:"key"`
	Topic   string `json:"topic"`
	Group   string `json:"group"`
	Count   int    `json:"count"`
	MaxSize int    `json:"max_size"`
}

func dlqKey(topic, group string) string {
	return topic + ":" + group
}

File: rust-heliosmq/src/dlq.rs
use std::sync::Arc;
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

use parking_lot::RwLock;

use crate::common::Message;

pub struct DLQManager {
    queues: RwLock<HashMap<String, DLQueue>>,
}

struct DLQueue {
    topic: String,
    group: String,
    messages: RwLock<Vec<Message>>,
    max_size: usize,
}

#[derive(Debug, Clone)]
pub struct DLQInfo {
    pub key: String,
    pub topic: String,
    pub group: String,
    pub count: usize,
    pub max_size: usize,
}

impl DLQManager {
    pub fn new() -> Self {
        Self {
            queues: RwLock::new(HashMap::new()),
        }
    }

    pub fn add_to_dlq(&self, msg: Message, group: &str, reason: &str) -> anyhow::Result<()> {
        let key = dlq_key(&msg.topic, group);
        
        let mut queues = self.queues.write();
        let queue = queues.entry(key.clone())
            .or_insert_with(|| DLQueue {
                topic: msg.topic.clone(),
                group: group.to_string(),
                messages: RwLock::new(Vec::new()),
                max_size: 10000,
            });

        let mut messages = queue.messages.write();
        
        if messages.len() >= queue.max_size {
            messages.remove(0);
        }

        let mut msg = msg;
        if msg.headers.is_empty() {
            msg.headers = HashMap::new();
        }
        msg.headers.insert("x-dlq-reason".to_string(), reason.to_string());
        msg.headers.insert("x-dlq-original-topic".to_string(), msg.topic.clone());
        msg.headers.insert("x-dlq-timestamp".to_string(), 
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_millis()
                .to_string()
        );

        messages.push(msg);

        Ok(())
    }

    pub fn get_dlq_messages(&self, topic: &str, group: &str, offset: usize, limit: usize) -> Option<(Vec<Message>, usize)> {
        let key = dlq_key(topic, group);
        let queues = self.queues.read();
        
        queues.get(&key).map(|queue| {
            let messages = queue.messages.read();
            let total = messages.len();
            
            if offset >= total {
                return (Vec::new(), total);
            }

            let end = (offset + limit).min(total);
            (messages[offset..end].to_vec(), total)
        })
    }

    pub fn retry_message(&self, topic: &str, group: &str, message_id: &str) -> anyhow::Result<Option<Message>> {
        let key = dlq_key(topic, group);
        let queues = self.queues.read();
        
        if let Some(queue) = queues.get(&key) {
            let mut messages = queue.messages.write();
            if let Some(pos) = messages.iter().position(|m| m.message_id == message_id) {
                let mut msg = messages.remove(pos);
                msg.deliver_at = 0;
                msg.retry_count += 1;
                return Ok(Some(msg));
            }
        }

        Ok(None)
    }

    pub fn list_dlqs(&self) -> Vec<DLQInfo> {
        let queues = self.queues.read();
        queues.iter().map(|(key, queue)| {
            DLQInfo {
                key: key.clone(),
                topic: queue.topic.clone(),
                group: queue.group.clone(),
                count: queue.messages.read().len(),
                max_size: queue.max_size,
            }
        }).collect()
    }
}

fn dlq_key(topic: &str, group: &str) -> String {
    format!("{}:{}", topic, group)
}
